src/main/java/hudson/plugins/groovy/WithGroovyStep.java

```diff
@@ -0,0 +1,244 @@
+import groovy.lang.GroovyObject;
+import groovy.lang.Writable;
+import hudson.AbortException;
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Util;
+import hudson.model.JDK;
+import hudson.model.Node;
+import hudson.model.TaskListener;
+import hudson.remoting.Channel;
+import hudson.remoting.ClassFilter;
+import hudson.remoting.ObjectInputStreamEx;
+import hudson.remoting.VirtualChannel;
+import hudson.remoting.Which;
+import hudson.slaves.WorkspaceList;
+import hudson.util.ListBoxModel;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import jenkins.model.Jenkins;
+import jenkins.security.MasterToSlaveCallable;
+import org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback;
+import org.jenkinsci.plugins.workflow.steps.EnvironmentExpander;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+/**
+ * Makes {@code groovy} in {@code PATH} be bound to an appropriate executable.
+ */
+public class WithGroovyStep extends Step {
+
+    private String tool;
+    private String jdk;
+    private Object input;
+
+    @DataBoundConstructor
+    public WithGroovyStep() {}
+
+    public String getTool() {
+        return tool;
+    }
+
+    @DataBoundSetter
+    public void setTool(String tool) {
+        this.tool = Util.fixEmpty(tool);
+    }
+
+    public String getJdk() {
+        return jdk;
+    }
+
+    @DataBoundSetter
+    public void setJdk(String jdk) {
+        this.jdk = Util.fixEmpty(jdk);
+    }
+
+    public Object getInput() {
+        return input;
+    }
+
+    @DataBoundSetter
+    public void setInput(Object input) {
+        this.input = input;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(context, this);
+    }
+
+    private static class Execution extends StepExecution {
+
+        private final transient WithGroovyStep step;
+
+        Execution(StepContext context, WithGroovyStep step) {
+            super(context);
+            this.step = step;
+        }
+
+        @Override
+        public boolean start() throws Exception {
+            FilePath base = WorkspaceList.tempDir(getContext().get(FilePath.class));
+            base.mkdirs();
+            FilePath tmp = base.createTempDir("jenkins-groovy-step", "");
+            Map<String, String> env = new HashMap<>();
+            // TODO these Remoting calls can block the CPS VM thread; need a SynchronousNonBlockingStepExecution analogue that applies only to the setup phase of a block-scoped step:
```

Comment:
This is not exactly ideal, but it's not grounds for blocking this either.

Reply:
The lack of timeouts here does make me twitch a bit too -- we could potentially hang the CPS thread indefinitely due to nonresponsive remoting.

Reply:
Okay, that's less of a concern then.

Reply:
Actually this API _does_ now exist.